common/Storyboarding/OsbSpriteWriter.cs

```diff
@@ -0,0 +1,306 @@
+        public static void AddCommand(this OsbSprite sprite, ICommand command)
+        {
+            if (command is ColorCommand colorCommand)
+                sprite.Color(colorCommand.Easing, colorCommand.StartTime, colorCommand.EndTime, colorCommand.StartValue, colorCommand.EndValue);
+            else if (command is FadeCommand fadeCommand)
+                sprite.Fade(fadeCommand.Easing, fadeCommand.StartTime, fadeCommand.EndTime, fadeCommand.StartValue, fadeCommand.EndValue);
+            else if (command is ScaleCommand scaleCommand)
+                sprite.Scale(scaleCommand.Easing, scaleCommand.StartTime, scaleCommand.EndTime, scaleCommand.StartValue, scaleCommand.EndValue);
+            else if (command is VScaleCommand vScaleCommand)
+                sprite.ScaleVec(vScaleCommand.Easing, vScaleCommand.StartTime, vScaleCommand.EndTime, vScaleCommand.StartValue, vScaleCommand.EndValue);
+            else if (command is ParameterCommand parameterCommand)
+                sprite.Parameter(parameterCommand.Easing, parameterCommand.StartTime, parameterCommand.EndTime, parameterCommand.StartValue);
+            else if (command is MoveCommand moveCommand)
+                sprite.Move(moveCommand.Easing, moveCommand.StartTime, moveCommand.EndTime, moveCommand.StartValue, moveCommand.EndValue);
+            else if (command is MoveXCommand moveXCommand)
+                sprite.MoveX(moveXCommand.Easing, moveXCommand.StartTime, moveXCommand.EndTime, moveXCommand.StartValue, moveXCommand.EndValue);
+            else if (command is MoveYCommand moveYCommand)
+                sprite.MoveY(moveYCommand.Easing, moveYCommand.StartTime, moveYCommand.EndTime, moveYCommand.StartValue, moveYCommand.EndValue);
+            else if (command is RotateCommand rotateCommand)
+                sprite.Rotate(rotateCommand.Easing, rotateCommand.StartTime, rotateCommand.EndTime, rotateCommand.StartValue, rotateCommand.EndValue);
+            else if (command is LoopCommand loopCommand)
+            {
+                sprite.StartLoopGroup(loopCommand.StartTime, loopCommand.LoopCount);
+                foreach (var cmd in loopCommand.Commands)
+                    AddCommand(sprite, cmd);
+                sprite.EndGroup();
+            }
+            else if (command is TriggerCommand triggerCommand)
+            {
+                sprite.StartTriggerGroup(triggerCommand.TriggerName, triggerCommand.StartTime, triggerCommand.EndTime, triggerCommand.Group);
+                foreach (var cmd in triggerCommand.Commands)
+                    AddCommand(sprite, cmd);
+                sprite.EndGroup();
+            }
+        }
+
+        public static bool IsFragmentable(this ICommand command)
+        {
+            if (command is ParameterCommand)
+                return true;
+
+            if (command.StartTime == command.EndTime)
+                return true;
+
+            var type = command.GetType();
+            var easingProp = type.GetProperty("Easing");

```

Comment:
That's evil af, either find an existing interface that has what you need or add make one, then `if (command is IHasEasingCommand easingCommand) somethingsomething(easingCommand.Easing)`

Reply:
That's a great idea. I tried to wrap my head around casting to generic interfaces before I resorted to reflection here, didn't think about moving the Easing property to a separate interface.

I moved both methods that used reflection to a new interface and implemented it for Command/CommandGroup respectively. This also made me realise, that the previous use of reflection would fail for compound commands. Fragmenting loops is a missing feature in this commit.